### PR TITLE
feature/548_add_column_mode_mongo_sink

### PR DESCRIPTION
**Pull request by xdelox**
Implements issue #548

### CODE REVIEW

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -0,0 +1 @@
+Issue #548 : added functionality: persist attribute persistence in column using mongo-sink
```
Comment (frbattid) [inline CHANGES_NEXT_RELEASE]: This must be:

    - [FEATURE] Column-like persistence in OrionMongoSink (#548)

Thanks! :)

File: src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java [hunk 2]
```diff
@@ -200,7 +201,8 @@ public void insertContextDataAggregated(String dbName, String collectionName, lo
      * @param attrValue
      * @param resolution
      */
-    private void insertContextDataAggregatedForResoultion(String dbName, String collectionName, GregorianCalendar calendar,
+    private void insertContextDataAggregatedForResoultion(String dbName, String collectionName,
+                                                          GregorianCalendar calendar,
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Is this simply a line feed?

File: src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java [hunk 3]
```diff
@@ -365,7 +366,7 @@ private BasicDBList buildPrepopulatedPoints(Resolution resolution) {
         return prepopulatedData;
     } // buildPrepopulatedPoints
     
-    /**
+  /**
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Please, add the white space again ;)

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoBaseSink.java [hunk 4]
```diff
@@ -46,6 +46,8 @@
     protected String collectionPrefix;
     protected boolean shouldHash;
     protected MongoBackend backend;
+    protected boolean rowAttrPersistence;
+
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoBaseSink.java]: This new line seems to be unncessary.

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java [hunk 5]
```diff
@@ -80,6 +81,11 @@ void persist(Map<String, String> eventHeaders, NotifyContextRequest notification
         } // if
         
         // iterate on the contextResponses
+
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java]: These four lines seem to be unncessary.

File: src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java [hunk 6]
```diff
@@ -192,7 +193,7 @@ public void insertContextDataAggregated(String dbName, String collectionName, lo
      * (row-like mode).
      * @param dbName
      * @param collectionName
-     * @param recvTimeTs
+     * @param calendar
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Thanks for fixing this! NTC

File: src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java [hunk 7]
```diff
@@ -375,9 +376,52 @@ private BasicDBList buildPrepopulatedPoints(Resolution resolution) {
...
+            case COLLECTIONPERENTITY:
+                it = attrs.entrySet().iterator();
+                while (it.hasNext()) {
+                    Map.Entry pair = (Map.Entry) it.next();
+                    String attrName = (String) pair.getKey();
+                    String attrValue = (String) pair.getValue();
+                    doc.append(attrName, attrValue);
+                }
+                break;
+            case COLLECTIONPERATTRIBUTE:
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: In fact, this data model has no sense when persisting the data in column-like mode. In this case I think a warning log must be written and nothing else must be done regarding the persistence.

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java [hunk 8]
```diff
@@ -122,13 +131,39 @@ void persist(Map<String, String> eventHeaders, NotifyContextRequest notification
...
-                        entityId, entityType, attrName, attrType, attrValue, attrMetadata);
+                if (this.rowAttrPersistence) {
+                    LOGGER.info("[" + this.getName() + "] Persisting data at OrionMongoSink. Database: "
+                            + dbName + ", Collection: " + collectionName + ", Data: " + recvTimeTs.longValue() / 1000L
+                            + "," + recvTime + "," + entityId + "," + entityType + ","
+                            + attrName + "," + attrType + "," + attrValue + "," + attrMetadata);
+                    this.backend.insertContextDataRaw(
+                            dbName, collectionName, recvTimeTs.longValue() / 1000L, recvTime,
+                            entityId, entityType, attrName, attrType, attrValue, attrMetadata);
+                } else if (this.dataModel == DataModel.COLLECTIONPERATTRIBUTE) {
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java]: If you see above, for me the COLLECTIONPERATTRIBUTE data model has no sense in the column-like persistence. Thus, this "else if" branch should probably removed.

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java [hunk 9]
```diff
@@ -122,13 +131,39 @@ void persist(Map<String, String> eventHeaders, NotifyContextRequest notification
...
+                    singleMds.put(attrName + "_md", attrMetadata);
+                    this.backend.insertContextDataRaw(
+                            dbName, collectionName, recvTimeTs.longValue() / 1000L, recvTime,
+                            entityId, entityType, singleAttrs, singleMds);
+                } else {
+                    attrs.put(attrName, attrValue);
+                    mds.put(attrName + "_md", attrMetadata);
+                }
             } // for
+            if (!this.rowAttrPersistence && this.dataModel != DataModel.COLLECTIONPERATTRIBUTE) {
```
Comment (frbattid) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java]: Yourselves have discovered the COLLECTIONPERATRIBUTE has no sense in column-like mode. In this case, a warn must be traced and nothis else must be done form the persistence point of view.

**Comment by frbattid** [comment]
First of all, many thanks for the PR! Second, apologies for delaying the review so much time.

Most of the code is perfect for me, simply check mu comments.

The only relevant issue is the usage of the COLLECTIONPERENTITY data model with the column-like attribute persistence. I've been thinking the best is to warn the user at the `configure` method when checking both parameters are configured at the same time; and exiting Cygnus with error. What do you think?

**Comment by xdelox** [comment]
I'm totally agree with you about COLLECTIONPERENTIRY data model.
So, what I've to do? Does I adjust the code according to your comments and doing another pull request?
I apologise for the trivial question, but I'm a newbie to GitHub (and Git). :)

File: src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java (hunk 2 above)
Comment (xdelox) [inline src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Yes. According to the max line length rule (120 characters) specified in the checkstyle guidelines, I splitted the existing row, in order to pass the check (even if  the method don't affect my feature).

**Comment by frbattid** [comment]
Hi! Simply perform the modifications at the same branch you are trying to merge (i.e. `xdelox:master`) and add more commits to it; those new commits will automatically appear in this PR, so no new PR has to be done ;)

File: src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java (hunk 3 above)
Comment (xdelox) [inline src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Done!

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoBaseSink.java (hunk 4 above)
Comment (xdelox) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoBaseSink.java]: Removed

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java (hunk 5 above)
Comment (xdelox) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java]: Removed

File: src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java (hunk 6 above)
Comment (xdelox) [inline src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: you're welcome! ;)

File: src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java (hunk 7 above)
Comment (xdelox) [inline src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java]: Done

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java (hunk 8 above)
Comment (xdelox) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java]: Done

File: src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java (hunk 9 above)
Comment (xdelox) [inline src/main/java/com/telefonica/iot/cygnus/sinks/OrionMongoSink.java]: Done, like other changes you suggested. 
Please, don't hesitate to contact me if something is wrong or inaccurate.

**Comment by frbattid** [comment]
The code seems OK for me. Nevertheless, before LGTM'ing the PR the documentation must be updated. Specifically:

* https://github.com/telefonicaid/fiware-cygnus/blob/master/doc/design/OrionMongoSink.md must describe the `attr_persistence` parameter, an example when using the column mode is required and the mapping from Flume events to MongoDB data structures must consider both methods (have a look about how this is done in https://github.com/telefonicaid/fiware-cygnus/blob/master/doc/design/OrionCKANSink.md).
* https://github.com/telefonicaid/fiware-cygnus/blob/master/README.md must contain updated examples when using the column mode (in fact, this is a copy&paste from the above file).

I'm also missing the new `attr_persistence` parameter in the `conf/agent.conf.template` and the README configuration section.

We are close to finish!

**Comment by frbattid** [comment]
Any update, @xdelox ? I've extended the milestone deadline until November 4th (included).

**Comment by xdelox** [comment]
Yes, sorry for the delay.... I'm working now in order to close the issue within this evening. ;)

**Comment by frbattid** [comment]
It sounds good! :)

**Comment by xdelox** [comment]
It seems to be completed... I've taken the liberty of adding information about _data_model_ parameter, tighly coupled with _data_column_, precedently not documented.

That's ok?

**Comment by frbattid** [comment]
I'm affraind I'm not getting you... what `data_column` parameter? :)

**Comment by xdelox** [comment]
Damn! Too much things in my head... I meant `<attr_persistence>` :dizzy_face:

File: README.md [hunk 10]
```diff
@@ -458,15 +459,82 @@ Assuming `mongo_username=myuser` and `should_hash=false` as configuration parame
...
+    sth_/4wheels
+    system.indexes
+    > db['sth_/4wheels'].find()
+    { "_id" : ObjectId("5534d143fa701f0be751db86"), "recvTimeTs": "1402409899391", "recvTime" : "2015-04-20T12:13:22.41.412Z", "entityId" : "car1", "entityType" : "car", "speed" : "112.9", "oil_level" : "74.6" }
+    
+If `data_model=collection-per-attribute` and `attr_persistence=row` then `OrionMongoSink` will persist the data as:
+
+    $ mongo -u myuser -p
+    MongoDB shell version: 2.6.9
+
```
Comment (frbattid) [inline README.md]: I guess something is missing here :)

**Comment by frbattid** [comment]
Everything seems OK for me except for the comment I made you regarding a missing paragraph in the doc.

Finally, one more thing: <b>by merging this PR Telefónica will own the full code, including your contribution, and thus Telefónica will own the intelectual property of all the code</b>. Nevertheless, that does not avoid to put your Github id in the authors section of each *.java file. So you can add yourself to the authors list if you wish ;)

**Comment by xdelox** [comment]
You have guessed correctly, another time... 
Well, I release to Telefónica all intellectual properties.
Now, let me see my name on the cygnus's wall of fame (added author annotation) !  ;)

**Comment by frbattid** [comment]
Great, this definitely LGTM :)

Nevertheless, this PR cannot be merged until you fix the conflicts that would arise if your local branch is directly merged into my upstream develop. This is because there have been changes in the upstream develop that are not within your forked repo. In order to get those changes, you will have to [merge the upstream develop into your local branch](https://help.github.com/articles/merging-an-upstream-repository-into-your-fork/); this will do the conflicts appear in your local branch, they are fixed as explained [here](https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line/). Then, both branches will be synchronized except for the new code your are adding.

**Comment by xdelox** [comment]
Just done.

**Comment by frbattid** [comment]
Many thanks for your effort, @xdelox !!! Hope to see you in another PR :)

**Pull request merged**